drivers/sensor/dht20/dht20.c

```diff
@@ -0,0 +1,338 @@
+
+/*
+ * According to datasheet 7.4
+ * Reset register 0x1B, 0x1C and 0x1E
+ */
+#define DHT20_RESET_REGISTER_0 0x1B
+#define DHT20_RESET_REGISTER_1 0x1C
+#define DHT20_RESET_REGISTER_2 0x1E
+
+LOG_MODULE_REGISTER(DHT20, CONFIG_SENSOR_LOG_LEVEL);
+
+struct dht20_config {
+	struct i2c_dt_spec bus;
+};
+
+struct dht20_data {
+	uint32_t t_sample;
+	uint32_t rh_sample;
+};
+
+/**
+ * @brief Read status register
+ *
+ * @param dev Pointer to the sensor device
+ * @param[out] status Pointer to which the status will be stored
+ * @return 0 if successful
+ */
+static inline int read_status(const struct device *dev, uint8_t *status)
+{
+	const struct dht20_config *cfg = dev->config;
+	int rc;
+	uint8_t tx_buf[] = {DHT20_STATUS_REGISTER};
+	uint8_t rx_buf[1];
+
+	/* Write DHT20_STATUS_REGISTER then read to get status */
+	rc = i2c_write_dt(&cfg->bus, tx_buf, sizeof(tx_buf));
+	if (rc < 0) {
+		LOG_ERR("Failed to start measurement.");
+		return rc;
+	}
+
+	rc = i2c_read_dt(&cfg->bus, rx_buf, sizeof(rx_buf));
+	if (rc < 0) {
+		LOG_ERR("Failed to read data from device.");
+		return rc;
+	}
+
+	/* Retrieve status from rx_buf */
+	*status = rx_buf[0];
+	return rc;
+}
+
+static inline int reset_register(const struct device *dev, uint8_t reg)
+{
+	const struct dht20_config *cfg = dev->config;
+	int rc;
+	uint8_t tx_buf[] = {reg, 0, 0};
+	uint8_t rx_buf[3];
+
+	/* Write and then read 3 bytes from device */
+	rc = i2c_write_read_dt(&cfg->bus, tx_buf, sizeof(tx_buf), rx_buf, sizeof(rx_buf));
+	if (rc < 0) {
+		LOG_ERR("Failed to reset register.");
+		return rc;
+	}
+
+	/* Write register again, using values read earlier */
+	tx_buf[0] = DHT20_MASK_RESET_REGISTER | reg;
+	tx_buf[1] = rx_buf[1];
+	tx_buf[2] = rx_buf[2];
+	rc = i2c_write_dt(&cfg->bus, tx_buf, sizeof(tx_buf));
+	if (rc < 0) {
+		LOG_ERR("Failed to reset register.");
+		return rc;
+	}
+
+	return rc;
+}
+
+static inline int reset_sensor(const struct device *dev)
+{
+	int rc;
+	uint8_t status;
+
+	rc = read_status(dev, &status);
+	if (rc < 0) {
+		LOG_ERR("Failed to read status");
+		return rc;
+	}
+
+	if ((status & DHT20_STATUS_MASK_CHECK) != DHT20_STATUS_MASK_CHECK) {
+		/*
+		 * According to datasheet 7.4
+		 * Reset register 0x1B, 0x1C and 0x1E if status does not match expected value
+		 */
+		rc = reset_register(dev, DHT20_RESET_REGISTER_0);
+		if (rc < 0) {
+			return rc;
+		}
+		rc = reset_register(dev, DHT20_RESET_REGISTER_1);
+		if (rc < 0) {
+			return rc;
+		}
+		rc = reset_register(dev, DHT20_RESET_REGISTER_2);
+		if (rc < 0) {
+			return rc;
+		}
+		/* Wait 10ms after reset sequence */
+		k_msleep(10);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Compute CRC8 value for checkung measurement data validity
+ *
+ * @details DHT20 uses CRC8 algorithm with initial value 0xFF and
+ * polynom 1 + X^4 + X^5 + X^8
+ *
+ * @param ptr Pointer to the measurement data
+ * @param len Length of the measurement data
+ * @return uint8_t CRC8 value
+ */
+static uint8_t dht20_crc8(uint8_t *ptr, uint8_t len)
```

Comment:
Can't you use zephyr's `crc8()`?

Reply:
Yes you are right. I updated the code to use crc8, and also put a Kconfig to enable or disable the check